python/tvm/relay/op/contrib/arm_compute_lib.py

```diff
@@ -52,11 +56,15 @@ def partition_for_arm_compute_lib(mod, params=None, **opts):
         The module to run passes on.
     params : Optional[Dict[str, NDArray]]
         Constant input parameters.
+    offload_concat : Optional[bool]
+        Whether offload concatenate
 
     Returns
     -------
     ret : annotated and partitioned module.
     """
+    global offload_concat_
```

Comment:
Rather than using a global var, can we make use of `arm_compute_lib_pattern_table`? https://github.com/apache/tvm/blob/8a0472fd3e3138ff17391b449cc02de3ab2c4207/python/tvm/relay/op/contrib/arm_compute_lib.py#L131

i.e. register `concat` pattern if `offload_concat == `.

Reply:
Thanks for taking a look @masahi. Currently ACL has a mixture of registering operations using composite functions and `_register_external_op_helper` which adds the `target.arm_compute_lib` attribute, which is not ideal. I think if we wanted to do this for concat it would need to be registered using pattern table rather than the other mechanism.

In the future we should probably use the pattern table for all operations.

Reply:
Yeah, if we are adding a new op support, why not add it via pattern like https://github.com/apache/tvm/blob/8a0472fd3e3138ff17391b449cc02de3ab2c4207/python/tvm/relay/op/contrib/arm_compute_lib.py#L134

Then it is enable / disable a pattern based on some flags. @DzAvril

Reply:
Updated, please review. I'm not sure it is what you want because it seems a little weird since `arm_compute_lib_pattern_table` aims to register fusing patterns for ops and registering op attr is out of it.
BTW, this commit triggered a `Sanity Check` error:
> python/tvm/relay/op/contrib/arm_compute_lib.py:274: [W0612(unused-variable), arm_compute_lib_pattern_table.concatenate] Unused variable 'concatenate'

Reply:
Actually I don't what `@tvm.ir.register_op_attr("nn.conv2d", "target.arm_compute_lib")` thing is for... Why do you need both the pattern + `check` function at 
https://github.com/apache/tvm/blob/8a0472fd3e3138ff17391b449cc02de3ab2c4207/python/tvm/relay/op/contrib/arm_compute_lib.py#L134 
https://github.com/apache/tvm/blob/8a0472fd3e3138ff17391b449cc02de3ab2c4207/python/tvm/relay/op/contrib/arm_compute_lib.py#L223

and `register_op_attr` thing at 
https://github.com/apache/tvm/blob/8a0472fd3e3138ff17391b449cc02de3ab2c4207/python/tvm/relay/op/contrib/arm_compute_lib.py#L291-L292 
?

I never needed the latter in the BYOCs I worked on.

Reply:
>  I'm not sure it is what you want 

No this is not what I had in mind. What I would do is to add something like
`("arm_compute_lib.concatenate", concat_pattern(), check_concat)`
at https://github.com/apache/tvm/blob/8a0472fd3e3138ff17391b449cc02de3ab2c4207/python/tvm/relay/op/contrib/arm_compute_lib.py#L276

Reply:
> Actually I don't know what `@tvm.ir.register_op_attr("nn.conv2d", "target.arm_compute_lib")` thing is for... 

Gently ping @[lhutton1](https://github.com/lhutton1), can you help to explain this to masai?

Reply:
> Actually I don't know what `@tvm.ir.register_op_attr("nn.conv2d", "target.arm_compute_lib")`

Yeah I don't think we need to register conv2d like this either as it should have already been picked up by the pattern table. This way of registering operators predates the pattern table so I guess this was just overlooked. We can follow up in a separate PR to move all operators so that they are registered using the pattern table - which I believe is the way other BYOC's are going. Apologies for the confusion @DzAvril, like @masahi said, it would be best to register concatenate (and other operators added in the future) using the pattern table since this is the preferred mechanism.

Reply:
Sorry for missing some comments above. I have a question about the `concatenate_pattern`, see code below.
```python
    def concatenate_pattern():
        """Create an concatenate pattern from equivalent relay operators.

        Returns
        -------
        pattern : dataflow_pattern.AltPattern
            Denotes the concatenate pattern.
        """
        pattern = is_op("concatenate")(wildcard())
        return pattern
```
As the input of `concatenate` is a tuple of tensors, what the pattern of the parameters should be like? `wildcard()` in the code block is not working.

Reply:
can you try `is_tuple(None)`? See https://github.com/apache/tvm/pull/7754